### DISCUSSION TITLE
Make `mdns::Event` `Clone`?

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3593] [Ideas]
Why not considering events from other protocols, or behaviours to be exact, are `Clone`? I mean users can construct an iterator themselves from the vector.

**Comment by thomaseizinger** [comment]
No objection on first thought, are you happy to send a PR?

**Reply by drHuangMHT** [reply]
Of course! But it will be a breaking change so I'm kind of worried about that.

**Reply by thomaseizinger** [reply]
Adding a trait implementation should not be a breaking change. Why do you think it is breaking?

**Reply by drHuangMHT** [reply]
I'm wondering why we are passing iterators here instead of the actual vector. Is there any benefit I don't see?

**Reply by thomaseizinger** [reply]
It predates my involvement with the library so I can't say for sure but `smallvec` esp. has seen a lot of use within `rust-libp2p`. I am not particularly attached to keep it. I do care about performance but I'd be surprised if heap allocations are our bottleneck anywhere.

I'd value ease of use and maintainability over unbenchmarked performance improvements. This is also inline with our recently added coding guidelines: https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#no-premature-optimizations

Removing those iterators and replacing them with `Vec` or something else will be a breaking change. I do care about not making unnecessary breaking changes so I'd stick with the iterator interface for now.

We should be able to make the event clone without changing that though.

**Reply by drHuangMHT** [reply]
It's totally doable and understandable. I used to be crazy about performance even when I just started programing until it almost brought my work to a halt and the guys at `tokio` discord channel really taught me a lesson. The reason why is that I'm really mad about the IM app I'm forced to use because of the insane user base it has. It tastes like s-, they even bundled an Unreal Engine for flashy avatars which I don't f-ing care. They don't even make it optional making the package as huge as 1 Gig. Moreover, they think they can 'split' the user base into two but in reality they are mostly overlapped making me, and millions of users to have two separate apps, each weigh about 5 Gigs, and easily over 10 Gig and even reach 30 Gigs when used heavily, while leaving basically no option to manage the taken-up space, in which stores data you have no access to, and it's your data, your chat history and the files you have received but "gone".
